File: HEAP (PRIORITY QUEUE)/heap_revision.py
class MinHeap:
    def __init__(self):
        self.heap = []

    def leftChildIndex(self,i):
        return (2*i + 1)
    def rightChildIndex(self,i):
        return (2*i + 2)
    def parentNodeIndex(self, i):
        return (i-1)//2
    
    def insert(self,val):
        self.heap.append(val);
        lastIndex = len(self.heap)-1
        self.heapify(lastIndex)
    def heapify(self, i):
        while i > 0:
            parentIndex = self.parentNodeIndex(i)
            if self.heap[parentIndex] > self.heap[i]:
                # swap
                self.heap[parentIndex],self.heap[i] = self.heap[i], self.heap[parentIndex]
                i = parentIndex
            else:
                break

heap = MinHeap()
heap.insert(10);
heap.insert(5);
heap.insert(20);
heap.insert(30);
heap.insert(2);
print(heap.heap)

File: HEAP (PRIORITY QUEUE)/heap_revision.js
class MinHeap {
    constructor() {
        // this.heap = [5, 10, 20, 30];
        this.heap = [];
    }

    leftChildIndex(i) {
        return (2 * i + 1);
    }

    rightChildIndex(i) {
        return (2 * i + 2);
    }

    parentNodeIndex(i) {
        return Math.floor((i - 1) / 2);
    }

    insert(val) {
        this.heap.push(val);
        let lastIndex = this.heap.length - 1;
        this.heapify(lastIndex);
    }

    heapify(i) {
        while (i > 0) {
            let parentIndex = this.parentNodeIndex(i);
            if (this.heap[parentIndex] > this.heap[i]) {
                // swap
                [this.heap[parentIndex], this.heap[i]] = [this.heap[i], this.heap[parentIndex]]
                i = parentIndex;
            } else {
                break;
            }
        }
    }
}

const heap = new MinHeap();
heap.insert(2)
console.log(heap)